File: visclient/FishSpecies.cpp
// -*-c++-*-

#include "FishSpecies.h"
#include "game.h"


namespace client
{

FishSpecies::FishSpecies(_FishSpecies* pointer)
{
    ptr = (void*) pointer;
}

int FishSpecies::id()
{
  return ((_FishSpecies*)ptr)->id;
}

char* FishSpecies::species()
{
  return ((_FishSpecies*)ptr)->species;
}

int FishSpecies::cost()
{
  return ((_FishSpecies*)ptr)->cost;
}

int FishSpecies::maxHealth()
{
  return ((_FishSpecies*)ptr)->maxHealth;
}

int FishSpecies::maxMovement()
{
  return ((_FishSpecies*)ptr)->maxMovement;
}

int FishSpecies::carryCap()
{
  return ((_FishSpecies*)ptr)->carryCap;
}

int FishSpecies::attackPower()
{
  return ((_FishSpecies*)ptr)->attackPower;
}

int FishSpecies::range()
{
  return ((_FishSpecies*)ptr)->range;
}

int FishSpecies::maxAttacks()
{
  return ((_FishSpecies*)ptr)->maxAttacks;
}

int FishSpecies::turnsTillAvailalbe()
{
  return ((_FishSpecies*)ptr)->turnsTillAvailalbe;
}

int FishSpecies::turnsTillUnavailable()
{
  return ((_FishSpecies*)ptr)->turnsTillUnavailable;
}


int FishSpecies::spawn(int x, int y)
{
  return fishSpeciesSpawn( (_FishSpecies*)ptr, x, y);
}



std::ostream& operator<<(std::ostream& stream,FishSpecies ob)
{
  stream << "id: " << ((_FishSpecies*)ob.ptr)->id  <<'\n';
  stream << "species: " << ((_FishSpecies*)ob.ptr)->species  <<'\n';
  stream << "cost: " << ((_FishSpecies*)ob.ptr)->cost  <<'\n';
  stream << "maxHealth: " << ((_FishSpecies*)ob.ptr)->maxHealth  <<'\n';
  stream << "maxMovement: " << ((_FishSpecies*)ob.ptr)->maxMovement  <<'\n';
  stream << "carryCap: " << ((_FishSpecies*)ob.ptr)->carryCap  <<'\n';
  stream << "attackPower: " << ((_FishSpecies*)ob.ptr)->attackPower  <<'\n';
  stream << "range: " << ((_FishSpecies*)ob.ptr)->range  <<'\n';
  stream << "maxAttacks: " << ((_FishSpecies*)ob.ptr)->maxAttacks  <<'\n';
  stream << "turnsTillAvailalbe: " << ((_FishSpecies*)ob.ptr)->turnsTillAvailalbe  <<'\n';
  stream << "turnsTillUnavailable: " << ((_FishSpecies*)ob.ptr)->turnsTillUnavailable  <<'\n';
  return stream;
}

}

File: client/csharp/FishSpecies.cs
using System;
using System.Runtime.InteropServices;


///
public class FishSpecies
{
  public IntPtr ptr;
  protected int ID;
  protected int iteration;

  public FishSpecies()
  {
  }

  public FishSpecies(IntPtr p)
  {
    ptr = p;
    ID = Client.fishSpeciesGetId(ptr);
    iteration = BaseAI.iteration;
  }

  public bool validify()
  {
    if(iteration == BaseAI.iteration) return true;
    for(int i = 0; i < BaseAI.fishSpeciess.Length; i++)
    {
      if(BaseAI.fishSpeciess[i].ID == ID)
      {
        ptr = BaseAI.fishSpeciess[i].ptr;
        iteration = BaseAI.iteration;
        return true;
      }
    }
    throw new ExistentialError();
  }

    //commands

  ///Have a new fish spawn and join the fight!
  public bool spawn(int x, int y)
  {
    validify();
    return (Client.fishSpeciesSpawn(ptr, x, y) == 0) ? false : true;
  }

    //getters


  ///Unique Identifier
  public int Id
  {
    get
    {
      validify();
      int value = Client.fishSpeciesGetId(ptr);
      return value;
    }
  }

  ///The fish species
  public string Species
  {
    get
    {
      validify();
      IntPtr value = Client.fishSpeciesGetSpecies(ptr);
      return Marshal.PtrToStringAuto(value);
    }
  }

  ///The amount of food it takes to raise this fish
  public int Cost
  {
    get
    {
      validify();
      int value = Client.fishSpeciesGetCost(ptr);
      return value;
    }
  }

  ///The maximum health of this fish
  public int MaxHealth
  {
    get
    {
      validify();
      int value = Client.fishSpeciesGetMaxHealth(ptr);
      return value;
    }
  }

  ///The maximum number of movements in a turn
  public int MaxMovement
  {
    get
    {
      validify();
      int value = Client.fishSpeciesGetMaxMovement(ptr);
      return value;
    }
  }

  ///The total weight the fish can carry
  public int CarryCap
  {
    get
    {
      validify();
      int value = Client.fishSpeciesGetCarryCap(ptr);
      return value;
    }
  }

  ///The power of the fish's attack
  public int AttackPower
  {
    get
    {
      validify();
      int value = Client.fishSpeciesGetAttackPower(ptr);
      return value;
    }
  }

  ///The attack arrange of the fish
  public int Range
  {
    get
    {
      validify();
      int value = Client.fishSpeciesGetRange(ptr);
      return value;
    }
  }

  ///Maximum number of times this unit can attack per turn
  public int MaxAttacks
  {
    get
    {
      validify();
      int value = Client.fishSpeciesGetMaxAttacks(ptr);
      return value;
    }
  }

  ///If this species is able to use stealth
  public int CanStealth
  {
    get
    {
      validify();
      int value = Client.fishSpeciesGetCanStealth(ptr);
      return value;
    }
  }

  ///How many turns until you can spawn this fish species
  public int TurnsTillAvailalbe
  {
    get
    {
      validify();
      int value = Client.fishSpeciesGetTurnsTillAvailalbe(ptr);
      return value;
    }
  }

  ///How many turns until you can no longer spawn this fish species
  public int TurnsTillUnavailable
  {
    get
    {
      validify();
      int value = Client.fishSpeciesGetTurnsTillUnavailable(ptr);
      return value;
    }
  }

}

